plugins/net_plugin/net_plugin.cpp

```diff
@@ -789,6 +858,108 @@ namespace eosio {
+                     ("peer", peer_name())("name", participant_name())("enabled", participating));
+            set_participating(participating);
+         }
+      }
+
+      std::string certificate_subject(ssl::verify_context &ctx) {
+         X509 *cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+         if (!cert)
+            return "";
+         
+         char buf[256];
+         //printing full subject string of rejected certificate
+         //openssl doesn't recommend this function for use due to lack of support of unicode and issues with '\'
+         //however there is no single function that can do the same.
+         //TODO: investigate usage of X509_NAME_print_ex , compare performance
+         //see http://openssl.6102.n7.nabble.com/openssl-org-1425-Request-make-X509-NAME-oneline-use-same-formatter-as-X509-NAME-print-ex-td33142.html
+         char* subject_str = X509_NAME_oneline(X509_get_subject_name(cert), buf, std::size(buf));
+
+         return {subject_str};
+      }
+
+      bool verify_certificate(bool preverified, ssl::verify_context &ctx) {
+         peer_dlog(this, "preverified: [${p}] certificate subject: ${s}", ("p", preverified)("s", certificate_subject(ctx)));
+         //certificate depth means number of certificate issuers verified current certificate
+         //openssl provides those one by one starting from root certificate
+         //we don't use CA certificate or intermidiate issuers, so skipping those
+         //we interested only in last certificate in the chain, i.e. the one that identifies client
+         auto depth = X509_STORE_CTX_get_error_depth(ctx.native_handle());
+         if (depth > 0) {
+            //preverified is true when certificate matches verification chain, provided via load_verify_file
+            return preverified;
+         }
+
+         //return pointer is managed by openssl
+         X509 *cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+         if (!cert) {
+            peer_dlog(this, "X509_STORE_CTX_get_current_cert returned null certificate");
+            return false;
+         }
+
+         //max subject size is 256 bytes
+         char buf[256];
+         //return pointer is managed by openssl
+         X509_NAME* subj = X509_get_subject_name(cert);
+         int length = X509_NAME_get_text_by_NID(subj, NID_organizationName, buf, std::size(buf));
+         
+         if (length > 0) {
+            std::string organization{buf, (size_t)length};
+            if (is_string_valid_name(organization)){
+               participant_name_ = account_name{organization};
+               //participant name will be later used in start_session to determine if
+               //participant is participating or not
+            }
+            else {
+               peer_dlog(this, "received unauthorized participant: ${s}", ("s", organization));
+            }
+         }
```

Comment:
This section of code appears not be related to verify cert. Can it be a separate function like get_organization?

Reply:
I pulled it into a method called extract_participant_name, but it still needs to be called in verify_cert. Is the real issue that we should call this method process_cert and then it is able to both verify the certificate and process any data it needs from it?

Reply:
process_cert describes what the function does better.